### PR TITLE
randomized tests

### PR DESCRIPTION
**Pull request by jw3126**

### CODE REVIEW

**Comment by TeroFrondelius** [comment]
I propose someone will try if same is true for the original C++ implementation. Original implementation wrapped the C++ code. Is that still runnable?

**Comment by jw3126** [comment]
I modified the wrapper to run with Cxx, see #20 . The test fails also with the cxx version.

**Comment by jw3126** [comment]
I found a 2d example with only 5 points, that goes wrong:
2d

**Comment by TeroFrondelius** [comment]
Nice example. Did you test this also with C++ version? Is this bug in Julia package or in the algorithm? Our internal software (legacy version) use the C++ version, which is a bit harder to test that's why I am interested. Also if this is a bug in the original C++ implementation it would be polite to let the original author to know it as well.

**Comment by jw3126** [comment]
Yeah the example is also broken with the Cxx version. You should be able to build and run the example on 0.6 automatically by checking out the cxx branch #20.

**Comment by jw3126** [comment]
Could you reproduce the cxx counter example? Just to make sure. Then we can inform the original author.

**Comment by TeroFrondelius** [comment]
After some frustration, because I first tried with my mac powerbook and gave up. Then tried with Windows 10 machine, which of course was a bad idea in the first place, no way to install Cxx.jl. Finally I dug out my linux laptop and had to install julia and all the packages. Cxx was difficult because it wanted to have a newer version of cmake that is available from apt-get, thus installing cmake couple of times. Still no success see: https://github.com/Keno/Cxx.jl/issues/373

I can reproduce the problems with the pure julia version:
```bash
tero@tero-Dell-System-XPS-L321X ~/.julia/v0.6/Miniball $ git fetch origin pull/18/head:problem
From https://github.com/JuliaFEM/Miniball.jl
 * [new ref]         refs/pull/18/head -> problem
tero@tero-Dell-System-XPS-L321X ~/.julia/v0.6/Miniball $ git checkout problem 
Previous HEAD position was 5b99493... Add stable docs tag
Switched to branch 'problem'
tero@tero-Dell-System-XPS-L321X ~/.julia/v0.6/Miniball $
```
```julia
julia> Pkg.test("Miniball")
INFO: Computing test dependencies for Miniball...
INFO: Installing JSON v0.17.2
INFO: Installing Lint v0.5.2
INFO: Installing Nullables v0.0.5
INFO: Testing Miniball
INFO: 5 test files found.
INFO: ----- Running tests from file test_2d_points.jl -----
INFO: 2D points clear!
INFO: ----- Testing file test_2d_points.jl completed in 2.46 seconds -----
INFO: ----- Running tests from file test_3d_points.jl -----
INFO: 3D points clear!
INFO: ----- Testing file test_3d_points.jl completed in 0.68 seconds -----
INFO: ----- Running tests from file test_broken.jl -----
INFO: ----- Testing file test_broken.jl completed in 0.54 seconds -----
INFO: ----- Running tests from file test_random.jl -----
INFO: ----- Testing file test_random.jl completed in 3.34 seconds -----
INFO: ----- Running tests from file test_simple.jl -----
INFO: ----- Testing file test_simple.jl completed in 6.22 seconds -----
Test Summary: | Pass  Broken  Total
Miniball.jl   | 3238      14   3252

Test statistics:
 ────────────────────────────────────────────────────────────────────────────
                                     Time                   Allocations      
                             ──────────────────────   ───────────────────────
      Tot / % measured:           15.3s / 85.2%            820MiB / 96.2%    

 Section             ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 test_simple.jl           1    6.22s  47.6%   6.22s    104MiB  13.2%   104MiB
 test_random.jl           1    3.34s  25.6%   3.34s    630MiB  79.9%   630MiB
 test_2d_points.jl        1    2.29s  17.5%   2.29s   45.7MiB  5.80%  45.7MiB
 test_3d_points.jl        1    676ms  5.18%   676ms   2.07MiB  0.26%  2.07MiB
 test_broken.jl           1    540ms  4.13%   540ms   6.93MiB  0.88%  6.93MiB
 ────────────────────────────────────────────────────────────────────────────
INFO: Miniball tests passed
INFO: Removing JSON v0.17.2
INFO: Removing Lint v0.5.2
INFO: Removing Nullables v0.0.5

julia> 
```

Any hints to install Cxx.jl and I can rerun the tests. @jw3126 can you add a test with those five 2D points, or did I miss it? Actually I don't need the test just the exact coordinates, it would be so much easier to request @ahojukka5 or someone else to test pure C++ version of our internal code.

**Comment by jw3126** [comment]
The example with 5 points in 2d and one with 6 points in 3d is included both in this branch and in the cxx branch. Look at the `test_broken.jl` file. 
Cxx is a bit fragile, I also often have issues with it. This time it first segfaulted, however after `Pkg.rm` removing and adding `Pkg.add` it it worked. I have Ubuntu 18.04,
julia 0.6.2, Cxx 0.2 and gcc 7.3.0 on this machine.

File: test/test_broken.jl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# This file is a part of JuliaFEM/MiniBall.jl.
+# License is GPL: see https://github.com/JuliaFEM/MiniBall.jl/blob/master/LICENSE.md
+
+using Base.Test
+using Miniball
+
+@testset "broken 5 points 2d" begin
+
+    pts = [-2.628921573512486897783446693210862576961517333984375000000000000000000000000000 4.345646543473985268235537660075351595878601074218750000000000000000000000000000; -7.783166805400283383420401150942780077457427978515625000000000000000000000000000e-01 -4.950583883652715400103261345066130161285400390625000000000000000000000000000000; -7.110427483983996488348111597588285803794860839843750000000000000000000000000000 9.078042051418130364837111301312688738107681274414062500000000000000000000000000e-01; -2.781077447672829272562466940144076943397521972656250000000000000000000000000000 -5.261543462775741808457041770452633500099182128906250000000000000000000000000000; -6.913805339049894804759333055699244141578674316406250000000000000000000000000000 1.448828316446829855834721456631086766719818115234375000000000000000000000000000]
```
Comment (jw3126) [inline test/test_broken.jl]: These are the five points!

**Comment by TeroFrondelius** [comment]
I tried upgrading. Now I have these: Linux Mint 17.3, julia 0.6.3 and gcc 7.3.0. Sorry no luck, still precompiling fails. 

```julia
julia> Pkg.rm("Cxx")
INFO: Removing Compat v0.68.0
INFO: Removing Cxx v0.2.0
INFO: Package database updated

julia> Pkg.add("Cxx")
INFO: Installing Compat v0.68.0
INFO: Installing Cxx v0.2.0
INFO: Building Cxx
writing path.jl file
Tuning for julia installation at /home/tero/Downloads/julia-d55cadc350/bin with sources possibly at /home/tero/Downloads/julia-d55cadc350/bin/../..
INFO: Building julia binary build
make: Nothing to be done for `all'.
INFO: Package database updated

julia> using Cxx
INFO: Precompiling module Cxx.
ERROR: LoadError: LoadError: could not load library "/home/tero/.julia/v0.6/Cxx/src/../deps/usr/lib/libcxxffi"
/home/tero/.julia/v0.6/Cxx/src/../deps/usr/lib/libcxxffi.so: undefined symbol: _ZTVN4llvm2cl3optISsLb0ENS0_6parserISsEEEE
Stacktrace:
 [1] dlopen(::String, ::UInt32) at ./libdl.jl:97
 [2] init_libcxxffi() at /home/tero/.julia/v0.6/Cxx/src/initialization.jl:20
 [3] include_from_node1(::String) at ./loading.jl:576
 [4] include(::String) at ./sysimg.jl:14
 [5] include_from_node1(::String) at ./loading.jl:576
 [6] include(::String) at ./sysimg.jl:14
 [7] anonymous at ./<missing>:2
while loading /home/tero/.julia/v0.6/Cxx/src/initialization.jl, in expression starting on line 22
while loading /home/tero/.julia/v0.6/Cxx/src/Cxx.jl, in expression starting on line 163
ERROR: Failed to precompile Cxx to /home/tero/.julia/lib/v0.6/Cxx.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:710
 [2] _require(::Symbol) at ./loading.jl:497
 [3] require(::Symbol) at ./loading.jl:405

julia> 
```

**Comment by TeroFrondelius** [comment]
I am not C++ expert, but I tried this https://gist.github.com/TeroFrondelius/092c1f99584f0d4420100d6b0f24a583

with this output: 
```bash
tero@tero-Dell-System-XPS-L321X ~/dev/miniball_problem $ ./a.out 
Center:
  -3.84606 -1.75088 
Squared radius:
  19.6492
Number of support points:
  2
Support point indices (numbers refer to the input order):
  4 1 
Relative error:
  0.966955
Suboptimality:
  0
Validity:
  possibly invalid
Computation time was 2.3e-05 seconds
tero@tero-Dell-System-XPS-L321X ~/dev/miniball_problem $ 
```
If this reproduces the error, you can make the bug report to the original Miniball.

**Comment by jw3126** [comment]
@TeroFrondelius I informed the Prof. Gaertner, with an email address that I found in the git log in cc. Tell me if the mail did not reach you.

**Comment by TeroFrondelius** [comment]
I found another set of bad points, see https://gist.github.com/TeroFrondelius/092c1f99584f0d4420100d6b0f24a583 there is also the julia code, which was used to find the bad set of points. C++ output:
```bash
tero@tero-Dell-System-XPS-L321X ~/dev/miniball_problem $ ./a.out 
Center:
  -0.527588 0.138745 
Squared radius:
  0.702401
Number of support points:
  2
Support point indices (numbers refer to the input order):
  3 1 
Relative error:
  2.15302
Suboptimality:
  0
Validity:
  possibly invalid
Computation time was 3e-05 seconds
tero@tero-Dell-System-XPS-L321X ~/dev/miniball_problem $ 
```
Center should be origo, because the points are selected from the unit circle.

**Comment by jw3126** [comment]
Cool, this is a very nice example!

**Comment by TeroFrondelius** [comment]
@jw3126 should I merge this pull request? My understanding is that this doesn't break anything.

**Comment by jw3126** [comment]
Yes you can merge it.

**Comment by ahojukka5** [comment]
Do you have code to make visualization of test? Interestingly, after refactoring code in PR #26, it looks that broken tests are broken meaning that they are actually working?  Does these make any sense with visual inspection:

```julia
julia> ball_radius
4.808532542524384

julia> ball.center
2-element Array{Float64,1}:
 -2.50089
 -0.461181
```

**Pull request merged**